File: snapshots/aes-bslice/aes.c
#include <inttypes.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>

static uint8_t sbox[256] = {
      (0x63), (0x7C), (0x77), (0x7B), (0xF2), (0x6B), (0x6F), (0xC5), (0x30), (0x01), (0x67), (0x2B), (0xFE), (0xD7), (0xAB), (0x76),
    (0xCA), (0x82), (0xC9), (0x7D), (0xFA), (0x59), (0x47), (0xF0), (0xAD), (0xD4), (0xA2), (0xAF), (0x9C), (0xA4), (0x72), (0xC0),
    (0xB7), (0xFD), (0x93), (0x26), (0x36), (0x3F), (0xF7), (0xCC), (0x34), (0xA5), (0xE5), (0xF1), (0x71), (0xD8), (0x31), (0x15),
    (0x04), (0xC7), (0x23), (0xC3), (0x18), (0x96), (0x05), (0x9A), (0x07), (0x12), (0x80), (0xE2), (0xEB), (0x27), (0xB2), (0x75),
    (0x09), (0x83), (0x2C), (0x1A), (0x1B), (0x6E), (0x5A), (0xA0), (0x52), (0x3B), (0xD6), (0xB3), (0x29), (0xE3), (0x2F), (0x84),
    (0x53), (0xD1), (0x00), (0xED), (0x20), (0xFC), (0xB1), (0x5B), (0x6A), (0xCB), (0xBE), (0x39), (0x4A), (0x4C), (0x58), (0xCF),
    (0xD0), (0xEF), (0xAA), (0xFB), (0x43), (0x4D), (0x33), (0x85), (0x45), (0xF9), (0x02), (0x7F), (0x50), (0x3C), (0x9F), (0xA8),
    (0x51), (0xA3), (0x40), (0x8F), (0x92), (0x9D), (0x38), (0xF5), (0xBC), (0xB6), (0xDA), (0x21), (0x10), (0xFF), (0xF3), (0xD2),
    (0xCD), (0x0C), (0x13), (0xEC), (0x5F), (0x97), (0x44), (0x17), (0xC4), (0xA7), (0x7E), (0x3D), (0x64), (0x5D), (0x19), (0x73),
    (0x60), (0x81), (0x4F), (0xDC), (0x22), (0x2A), (0x90), (0x88), (0x46), (0xEE), (0xB8), (0x14), (0xDE), (0x5E), (0x0B), (0xDB),
    (0xE0), (0x32), (0x3A), (0x0A), (0x49), (0x06), (0x24), (0x5C), (0xC2), (0xD3), (0xAC), (0x62), (0x91), (0x95), (0xE4), (0x79),
    (0xE7), (0xC8), (0x37), (0x6D), (0x8D), (0xD5), (0x4E), (0xA9), (0x6C), (0x56), (0xF4), (0xEA), (0x65), (0x7A), (0xAE), (0x08),
    (0xBA), (0x78), (0x25), (0x2E), (0x1C), (0xA6), (0xB4), (0xC6), (0xE8), (0xDD), (0x74), (0x1F), (0x4B), (0xBD), (0x8B), (0x8A),
    (0x70), (0x3E), (0xB5), (0x66), (0x48), (0x03), (0xF6), (0x0E), (0x61), (0x35), (0x57), (0xB9), (0x86), (0xC1), (0x1D), (0x9E),
    (0xE1), (0xF8), (0x98), (0x11), (0x69), (0xD9), (0x8E), (0x94), (0x9B), (0x1E), (0x87), (0xE9), (0xCE), (0x55), (0x28), (0xDF),
    (0x8C), (0xA1), (0x89), (0x0D), (0xBF), (0xE6), (0x42), (0x68), (0x41), (0x99), (0x2D), (0x0F), (0xB0), (0x54), (0xBB), (0x16)
};

typedef uint8_t* state_t;

 void subBytes(state_t st) {
  for (int i = 0; i < 16; i++)
    st[i] = sbox[st[i]];
}

 void shiftRow(int i, int shift, state_t st){
   uint8_t s0 = st[i + (4 * (shift % 4))];
   uint8_t s1 = st[i + (4 * ((shift + 1) % 4))];
   uint8_t s2 = st[i + (4 * ((shift + 2) % 4))];
   uint8_t s3 = st[i + (4 * ((shift + 3) % 4))];
   st[i] = s0;
   st[i+4] = s1;
   st[i+8] = s2;
   st[i+12] = s3;
}

 void shiftRows(state_t st) {
  shiftRow(1,1,st);
  shiftRow(2,2,st);
  shiftRow(3,3,st);
}

 uint8_t xtime(uint8_t x) {
  uint8_t x1 = x << 1;
  uint8_t x711b = (uint8_t)((x >> 7) & 1) * 0x1b;
  return x1 ^ x711b;
}

 void mixColumn(int c, state_t st) {
  int i0 = 4 * c;
  uint8_t s0 = st[i0];
  uint8_t s1 = st[i0+1];
  uint8_t s2 = st[i0+2];
  uint8_t s3 = st[i0+3];
  uint8_t tmp = s0 ^ s1 ^ s2 ^ s3;
  st[i0]   = s0 ^ tmp ^ xtime(s0 ^ s1);
  st[i0+1] = s1 ^ tmp ^ xtime(s1 ^ s2);
  st[i0+2] = s2 ^ tmp ^ xtime(s2 ^ s3);
  st[i0+3] = s3 ^ tmp ^ xtime(s3 ^ s0);
}

 void mixColumns(state_t st) {
  mixColumn(0,st);
  mixColumn(1,st);
  mixColumn(2,st);
  mixColumn(3,st);
}

 void addRoundKey(state_t st, state_t k) {
  for (int i = 0; i < 16; i++)
    st[i] ^= k[i];
}

 void aes_enc(state_t st, state_t k) {
  subBytes(st);
  shiftRows(st);
  mixColumns(st);
  addRoundKey(st,k);
}

 void aes_enc_last(state_t st, state_t k) {
  subBytes(st);
  shiftRows(st);
  addRoundKey(st,k);
}

void rounds(state_t st, uint8_t* key) {
  for (int i = 0; i < 9; i++)
    aes_enc(st,key+(16*i));
}

 void block_cipher(state_t out, uint8_t* key) {
  uint8_t* k0 = key;
  uint8_t* k = key + 16;
  uint8_t* kn = key + (16 * 10);
  addRoundKey(out,k0);
  rounds(out,k);
  aes_enc_last(out,kn);
}

 void rotate_word(uint8_t* w) {
  uint8_t w0 = w[0];
  w[0] = w[1];
  w[1] = w[2];
  w[2] = w[3];
  w[3] = w0;
}

 void sub_word(uint8_t* w) {
  w[0] = sbox[w[0]];
  w[1] = sbox[w[1]];
  w[2] = sbox[w[2]];
  w[3] = sbox[w[3]];
}

static uint8_t rcon[11] = {
  (0x8d), (0x01), (0x02), (0x04), (0x08), (0x10), (0x20), (0x40), (0x80), (0x1b), (0x36)
};

 void aes_keygen_assist(uint8_t* out, uint8_t rcon){
  rotate_word(out);
  sub_word(out);
  out[0] ^= rcon;
}

 void key_expansion_word(uint8_t* out, uint8_t* w0, uint8_t* w1, int i){
  memcpy(out,w1,4);
  if (i % 4 == 0) 
    aes_keygen_assist(out,rcon[i/4]);
  for (int i = 0; i < 4; i++)
    out[i] ^= w0[i];
}

 void key_expansion(uint8_t* out, state_t key) {
  memcpy(out,key,16);
  for (int i = 4; i < 44; i++) 
    key_expansion_word(out+(4*i),out+((4*i)-16),out+((4*i)-4),i);
 }

 void aes128_block(state_t out,uint8_t* kex, uint8_t* n, uint32_t c) {
  memcpy(out,n,12);
  out[12] = (uint8_t)(c >> 24);
  out[13] = (uint8_t)(c >> 16);
  out[14] = (uint8_t)(c >> 8);
  out[15] = (uint8_t)(c);
  block_cipher(out,kex);
}

void aes128_ctr(uint8_t* out, uint8_t* in, int in_len, uint8_t* k, uint8_t* n, uint32_t c) {
  uint8_t kex[16*11] = {0};
  key_expansion(kex,k);
  uint8_t kb[16] = {0};
  int blocks = in_len / 16;
  for (int i = 0; i < blocks; i++) {
    aes128_block(kb,kex,n,c+i);
    for (int j = 0; j < 16; j++)
      out[(16*i)+j] = in[(16*i)+j] ^ kb[j];
  }
  int rem = in_len % 16;
  if (rem > 0) {
      aes128_block(kb,kex,n,c+blocks);
      for (int j = 0; j < rem; j++)
	out[(16*blocks)+j] = in[(16*blocks)+j] ^ kb[j];
    }
}

void aes128_encrypt(uint8_t* out, uint8_t* in, int in_len, uint8_t* k, uint8_t* n, uint32_t c) {
  aes128_ctr(out,in,in_len,k,n,c);
}

void aes128_decrypt(uint8_t* out, uint8_t* in, int in_len, uint8_t* k, uint8_t* n, uint32_t c) {
  aes128_ctr(out,in,in_len,k,n,c);
}

File: snapshots/aes-bslice/aes.py
#!/usr/bin/python3

from speclib import *

blocksize = 16
block_t  = bytes_t(16)
subblock_t  = refine(vlbytes_t, lambda x: array.length(x) <= blocksize)

rowindex_t = range_t(0,4)
expindex_t = range_t(0,48)
word_t = bytes_t(4)
key_t    = bytes_t(32)
nonce_t  = bytes_t(12)


index_t  = range_t(0,16)
rotval_t = range_t(1,32)
state_t  = array_t(uint32_t,16)


sbox : array_t(uint8_t,256) = array([
    uint8(0x63), uint8(0x7C), uint8(0x77), uint8(0x7B), uint8(0xF2), uint8(0x6B), uint8(0x6F), uint8(0xC5), uint8(0x30), uint8(0x01), uint8(0x67), uint8(0x2B), uint8(0xFE), uint8(0xD7), uint8(0xAB), uint8(0x76),
    uint8(0xCA), uint8(0x82), uint8(0xC9), uint8(0x7D), uint8(0xFA), uint8(0x59), uint8(0x47), uint8(0xF0), uint8(0xAD), uint8(0xD4), uint8(0xA2), uint8(0xAF), uint8(0x9C), uint8(0xA4), uint8(0x72), uint8(0xC0),
    uint8(0xB7), uint8(0xFD), uint8(0x93), uint8(0x26), uint8(0x36), uint8(0x3F), uint8(0xF7), uint8(0xCC), uint8(0x34), uint8(0xA5), uint8(0xE5), uint8(0xF1), uint8(0x71), uint8(0xD8), uint8(0x31), uint8(0x15),
    uint8(0x04), uint8(0xC7), uint8(0x23), uint8(0xC3), uint8(0x18), uint8(0x96), uint8(0x05), uint8(0x9A), uint8(0x07), uint8(0x12), uint8(0x80), uint8(0xE2), uint8(0xEB), uint8(0x27), uint8(0xB2), uint8(0x75),
    uint8(0x09), uint8(0x83), uint8(0x2C), uint8(0x1A), uint8(0x1B), uint8(0x6E), uint8(0x5A), uint8(0xA0), uint8(0x52), uint8(0x3B), uint8(0xD6), uint8(0xB3), uint8(0x29), uint8(0xE3), uint8(0x2F), uint8(0x84),
    uint8(0x53), uint8(0xD1), uint8(0x00), uint8(0xED), uint8(0x20), uint8(0xFC), uint8(0xB1), uint8(0x5B), uint8(0x6A), uint8(0xCB), uint8(0xBE), uint8(0x39), uint8(0x4A), uint8(0x4C), uint8(0x58), uint8(0xCF),
    uint8(0xD0), uint8(0xEF), uint8(0xAA), uint8(0xFB), uint8(0x43), uint8(0x4D), uint8(0x33), uint8(0x85), uint8(0x45), uint8(0xF9), uint8(0x02), uint8(0x7F), uint8(0x50), uint8(0x3C), uint8(0x9F), uint8(0xA8),
    uint8(0x51), uint8(0xA3), uint8(0x40), uint8(0x8F), uint8(0x92), uint8(0x9D), uint8(0x38), uint8(0xF5), uint8(0xBC), uint8(0xB6), uint8(0xDA), uint8(0x21), uint8(0x10), uint8(0xFF), uint8(0xF3), uint8(0xD2),
    uint8(0xCD), uint8(0x0C), uint8(0x13), uint8(0xEC), uint8(0x5F), uint8(0x97), uint8(0x44), uint8(0x17), uint8(0xC4), uint8(0xA7), uint8(0x7E), uint8(0x3D), uint8(0x64), uint8(0x5D), uint8(0x19), uint8(0x73),
    uint8(0x60), uint8(0x81), uint8(0x4F), uint8(0xDC), uint8(0x22), uint8(0x2A), uint8(0x90), uint8(0x88), uint8(0x46), uint8(0xEE), uint8(0xB8), uint8(0x14), uint8(0xDE), uint8(0x5E), uint8(0x0B), uint8(0xDB),
    uint8(0xE0), uint8(0x32), uint8(0x3A), uint8(0x0A), uint8(0x49), uint8(0x06), uint8(0x24), uint8(0x5C), uint8(0xC2), uint8(0xD3), uint8(0xAC), uint8(0x62), uint8(0x91), uint8(0x95), uint8(0xE4), uint8(0x79),
    uint8(0xE7), uint8(0xC8), uint8(0x37), uint8(0x6D), uint8(0x8D), uint8(0xD5), uint8(0x4E), uint8(0xA9), uint8(0x6C), uint8(0x56), uint8(0xF4), uint8(0xEA), uint8(0x65), uint8(0x7A), uint8(0xAE), uint8(0x08),
    uint8(0xBA), uint8(0x78), uint8(0x25), uint8(0x2E), uint8(0x1C), uint8(0xA6), uint8(0xB4), uint8(0xC6), uint8(0xE8), uint8(0xDD), uint8(0x74), uint8(0x1F), uint8(0x4B), uint8(0xBD), uint8(0x8B), uint8(0x8A),
    uint8(0x70), uint8(0x3E), uint8(0xB5), uint8(0x66), uint8(0x48), uint8(0x03), uint8(0xF6), uint8(0x0E), uint8(0x61), uint8(0x35), uint8(0x57), uint8(0xB9), uint8(0x86), uint8(0xC1), uint8(0x1D), uint8(0x9E),
    uint8(0xE1), uint8(0xF8), uint8(0x98), uint8(0x11), uint8(0x69), uint8(0xD9), uint8(0x8E), uint8(0x94), uint8(0x9B), uint8(0x1E), uint8(0x87), uint8(0xE9), uint8(0xCE), uint8(0x55), uint8(0x28), uint8(0xDF),
    uint8(0x8C), uint8(0xA1), uint8(0x89), uint8(0x0D), uint8(0xBF), uint8(0xE6), uint8(0x42), uint8(0x68), uint8(0x41), uint8(0x99), uint8(0x2D), uint8(0x0F), uint8(0xB0), uint8(0x54), uint8(0xBB), uint8(0x16)
])


def subBytes(state:block_t) -> block_t:
    st = array.copy(state)
    for i in range(16):
        st[i] = sbox[uint8.to_int(state[i])]
    return st

def shiftRow(i:rowindex_t,shift:rowindex_t,state:block_t) -> block_t:
    out = array.copy(state)
    out[i] = state[i + (4 * (shift % 4))]
    out[i+4] = state[i + (4 * ((shift + 1) % 4))]
    out[i+8] = state[i + (4 * ((shift + 2) % 4))]
    out[i+12] = state[i + (4 * ((shift + 3) % 4))]
    return out

def shiftRows(state:block_t) -> block_t:
    state = shiftRow(1,1,state)
    state = shiftRow(2,2,state)
    state = shiftRow(3,3,state)
    return state

def xtime(x:uint8_t) -> uint8_t:
    x1 = x << 1
    x7 = x >> 7
    x71 = x7 & uint8(1)
    x711b = x71 * uint8(0x1b)
    return x1 ^ x711b

def mixColumn(c:rowindex_t,state:block_t) -> block_t:
    i0 = 4 * c
    s0 = state[i0]
    s1 = state[i0+1]
    s2 = state[i0+2]
    s3 = state[i0+3]
    st = array.copy(state)
    tmp = s0 ^ s1 ^ s2 ^ s3
    st[i0]   = s0 ^ tmp ^ (xtime (s0 ^ s1))
    st[i0+1] = s1 ^ tmp ^ (xtime (s1 ^ s2))
    st[i0+2] = s2 ^ tmp ^ (xtime (s2 ^ s3))
    st[i0+3] = s3 ^ tmp ^ (xtime (s3 ^ s0))
    return st

def mixColumns(state:block_t) -> block_t:
    state = mixColumn(0,state)
    state = mixColumn(1,state)
    state = mixColumn(2,state)
    state = mixColumn(3,state)
    return state

def addRoundKey(state:block_t,key:block_t) -> block_t:
    out = array.copy(state)
    for i in range(16):
        out[i] ^= key[i]
    return out

def aes_enc(state:block_t,round_key:block_t) -> block_t:
    state = subBytes(state)
    state = shiftRows(state)
    state = mixColumns(state)
    state = addRoundKey(state,round_key)
    return state

def aes_enc_last(state:block_t,round_key:block_t) -> block_t:
    state = subBytes(state)
    state = shiftRows(state)
    state = addRoundKey(state,round_key)
    return state

def rounds(state:block_t,key:bytes_t(9*16)) -> block_t:
    out = array.copy(state)
    for i in range(9):
        out = aes_enc(out,key[16*i:16*i+16])
    return out

def block_cipher(input:block_t,key:bytes_t(11*16)) -> block_t:
    state = array.copy(input)
    k0 = key[0:16]
    k  = key[16:10*16]
    kn = key[10*16:11*16]
    state = addRoundKey(state,k0)
    state = rounds(state,k)
    state = aes_enc_last(state,kn)
    return state

def rotate_word(w:word_t) -> word_t:
    out = array.copy(w)
    out[0] = w[1]
    out[1] = w[2]
    out[2] = w[3]
    out[3] = w[0]
    return out

def sub_word(w:word_t) -> word_t:
    out = array.copy(w)
    out[0] = sbox[uint8.to_int(w[0])]
    out[1] = sbox[uint8.to_int(w[1])]
    out[2] = sbox[uint8.to_int(w[2])]
    out[3] = sbox[uint8.to_int(w[3])]
    return out

rcon : bytes_t(11) = array([uint8(0x8d), uint8(0x01), uint8(0x02), uint8(0x04), uint8(0x08), uint8(0x10), uint8(0x20), uint8(0x40), uint8(0x80), uint8(0x1b), uint8(0x36)])

def aes_keygen_assist(w:word_t,rcon:uint8_t) -> word_t:
    k = rotate_word(w)
    k = sub_word(k)
    k[0] ^= rcon
    return k

def key_expansion_word(w0:word_t, w1:word_t, i:expindex_t) -> word_t:
    k = array.copy(w1)
    if i % 4 == 0:
        k = aes_keygen_assist(k,rcon[i//4])
    for i in range(4):
        k[i] ^= w0[i]
    return k

def key_expansion(key:block_t) -> bytes_t(11*16):
    key_ex = array.create(11*16,uint8(0))
    key_ex[0:16] = key
    for j in range(40):
        i = j + 4
        key_ex[4*i:4*i+4] = key_expansion_word(key_ex[4*i-16:4*i-12],key_ex[4*i-4:4*i],i)
    return key_ex

def aes128_block(k:key_t,n:nonce_t,c:uint32_t) -> block_t:
    input = array.create(16,uint8(0))
    input[0:12] = n
    input[12:16] = bytes.from_uint32_be(c)
    key_ex = key_expansion(k)
    out = block_cipher(input,key_ex)
    return out

def xor_block(block:subblock_t, keyblock:block_t) -> subblock_t:
    out = vlbytes.copy(block)
    for i in range(array.length(block)):
        out[i] ^= keyblock[i]
    return out

def aes128_counter_mode(key: key_t, nonce: nonce_t, counter: uint32_t, msg:vlbytes_t) -> vlbytes_t:
    blocks,last = vlarray.split_blocks(msg,blocksize)
    keyblock = array.create(blocksize,uint8(0))
    ctr = counter
    for i in range(array.length(blocks)):
        keyblock = aes128_block(key,nonce,ctr)
        blocks[i] = xor_block(blocks[i],keyblock)
        ctr += uint32(1)
    keyblock = aes128_block(key,nonce,ctr)
    last = xor_block(last,keyblock)
    return vlarray.concat_blocks(blocks,last)

def aes128_encrypt(key: key_t, nonce: nonce_t, counter: uint32_t, msg:vlbytes_t) -> vlbytes_t:
    return aes128_counter_mode(key,nonce,counter,msg)

def aes128_decrypt(key: key_t, nonce: nonce_t, counter: uint32_t, msg:vlbytes_t) -> vlbytes_t:
    return aes128_counter_mode(key,nonce,counter,msg)


